### PR TITLE
authelia: 4.37.5 -> 4.38.9

### LINKED ISSUE #296502: Update request: authelia 4.37.5 → 4.38.2
- Package name: `authelia`
- Latest released version: `4.38.2`

- Current version on the unstable channel: `4.37.5`
- Current version on the stable/release channel: `4.37.5`

- [x] Checked the [nixpkgs pull requests](https://github.com/NixOS/nixpkgs/pulls)

**Notify maintainers**

@06kellyjac @RaitoBezarius @dit7ya 

-----

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by nicomem**
## Description of changes

Closes #296502

### Package Update
Updated manually using the `update.sh` script (but as said in #296502, the nixpkgs-update bot will probably not be able to automatically update it).

The 4.38.0 version has some big changes and some configuration depreciation, but simply by keeping my old configuration, everything seems to work fine (apart from the depreciation warnings).

Changelogs:
https://github.com/authelia/authelia/releases/tag/v4.38.0
https://github.com/authelia/authelia/releases/tag/v4.38.1
https://github.com/authelia/authelia/releases/tag/v4.38.2
https://github.com/authelia/authelia/releases/tag/v4.38.3
https://github.com/authelia/authelia/releases/tag/v4.38.4
https://github.com/authelia/authelia/releases/tag/v4.38.5
https://github.com/authelia/authelia/releases/tag/v4.38.6
https://github.com/authelia/authelia/releases/tag/v4.38.7
https://github.com/authelia/authelia/releases/tag/v4.38.8
https://github.com/authelia/authelia/releases/tag/v4.38.9

#### Removals
- Removed options `services.authelia.instances.<name>.settings.server.{host,port}`, and added `services.authelia.instances.<name>.settings.server.address`.

If any of the former options are used, an assertion will be triggered.

#### Other changes

- `services.authelia.instances.<name>.secrets.jwtSecretFile` now sets the `AUTHELIA_IDENTITY_VALIDATION_RESET_PASSWORD_JWT_SECRET_FILE` environment variable instead of `AUTHELIA_JWT_SECRET_FILE`
- `services.authelia.instances.<name>.secrets.oidcIssuerPrivateKeyFile` now creates a new YAML config file to set `identity_providers.oidc.key` using [templates](https://www.authelia.com/reference/guides/templating/)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by nicomem** [comment]
Notify maintainers : @06kellyjac @RaitoBezarius @dit7ya

**Comment by GGG-KILLER** [comment]
Gave a try setting this up and the module will need reworking for people who want to take advantage of unix sockets

**Comment by b12f** [comment]
The service config will have to be updated. Using `server.address` , the new and recommended way to set the listening address, crashes Authelia. This is because the [nix options explicitly declare](https://github.com/NixOS/nixpkgs/blob/nixos-23.11/nixos/modules/services/security/authelia.nix#L164) `server.host` and `server.port`, and declaring both is invalid.

**Comment by b12f** [comment]
To make life simple on yourself, I would just deprecate the config option and show the user where to find the new one. You can use `lib.mkRemovedModuleOption` for this: https://github.com/NixOS/nixpkgs/blob/master/lib/modules.nix#L1091

**Comment by nicomem** [comment]
`lib.mkRemovedModuleOption` did not work for the same reasons as `lib.mkMergedOptionModule`, so I instead added an assertion checking if either option was set

**Comment by nicomem** [comment]
Updated again the module due to other options having problems with the new settings:
- `secrets.jwtSecretFile` was mapped to the `AUTHELIA_JWT_SECRET_FILE` env variable
  - It is now mapped to `AUTHELIA_IDENTITY_VALIDATION_RESET_PASSWORD_JWT_SECRET_FILE` (the newer setting)

- `secrets.oidcIssuerPrivateKeyFile` has been removed
  - Authelia now does not allow setting this using env variable, but by setting `identity_providers.oidc.jwks.key` (`jwks` being an array)
  - Authelia advises using a file filter as such:
    ```yaml
    identity_providers:
      oidc:
        jwks:
          - key: {{ secret "/config/jwks/rsa.2048.pem" | mindent 10 "|" | msquote }}
    ```
  - However, this way of doing does not seem to work with the Nix -> YAML conversion
  - So the more secure way I found of setting this is by using `services.authelia.<name>.settingsFiles`, and using a file that is only readable by the authelia user (for example, in my conf, I use [agenix](https://github.com/ryantm/agenix))

Updated the manual & added an assertion in the same way as the previously removed options

**Comment by nicomem** [comment]
Added a commit to add myself as a maintainer.

Re-pinging the maintainers : @06kellyjac @dit7ya

In case the maintainers still do not respond, also adding @AndersonTorres who merged the init PR (#180469), and @marsam who merged the last authelia PR (#266455)

**Comment by alapshin** [comment]
Looks good to me. Thank you for your work.

I have upgraded my self-hosted setup and everything seems to be working fine after adjustments for removed options.

Regarding replacement of `oidcIssuerPrivateKeyFile`

> Updated again the module due to other options having problems with the new settings:
> 
>     * `secrets.oidcIssuerPrivateKeyFile` has been removed
>       
>       * Authelia now does not allow setting this using env variable, but by setting `identity_providers.oidc.jwks.key` (`jwks` being an array)
>       * Authelia advises using a file filter as such:
>         ```yaml
>         identity_providers:
>           oidc:
>             jwks:
>               - key: {{ secret "/config/jwks/rsa.2048.pem" | mindent 10 "|" | msquote }}
>         ```
>       * However, this way of doing does not seem to work with the Nix -> YAML conversion
>       * So the most secure way I found of setting this is by using `services.authelia.<name>.settingsFiles`, and using a file that is only readable by the authelia user (for example, in my conf, I use [agenix](https://github.com/ryantm/agenix))

There are 2 issues for using approach from Authelia docs.

First, to enable template support in config files one have to launch Authelia with env variable `X_AUTHELIA_CONFIG_FILTERS=template`, see https://www.authelia.com/reference/guides/templating/#enable-templating

Second, when trying to generate this YAML section using nix
```
identity_providers = {
    oidc = {
        jwks = [
            {
                key_id = "main";
                key = ''{{ secret "/run/secrets/authelia/jwk_rsa_key.pem" | mindent 10 "|" | msquote }}'';
            }
        ];
    };
};
```

The resulting YAML config looks like this: closing braces for some reason are placed on a new line. Authelia doesn't expand the template and instead tries to read the key directly.
```
jwks:
    - key: '{{ secret "/run/secrets/authelia/jwk_rsa_key.pem" | mindent 10 "|" | msquote
        }}'
```

My solution is similar to yours.

I set env variable `X_AUTHELIA_CONFIG_FILTERS=template` in systemd service config.  Then use additional config and load it via `settingsFiles` option
```
identity_providers:
  oidc:
    jwks:
      - key: {{ secret "/run/secrets/authelia/jwk_rsa_key.pem" | mindent 10 "|" | msquote }}
```

Where `/run/secrets/authelia/jwk_rsa_key.pem` is a secret file managed by sops-nix.

File: nixos/doc/manual/release-notes/rl-2405.section.md
Comment (SuperSandro2000) [inline nixos/doc/manual/release-notes/rl-2405.section.md:42]:
```suggestion
  This release also depreciates some configuration keys, which are likely to be removed in future version 5.0, but they are still supported and expected to be working in the current version.
```

**Comment by SuperSandro2000** [comment]
> Tried to see if we could avoid vendoring the `package-lock.json` file, as the Authelia repo has a `pnpm-lock.yaml`, but it seems like there is not yet the required tooling for that (see #231513)

There is now 🎉

**Comment by nicomem** [comment]
Changes:
- Use the upstream `pnpm-lock.json` instead of creating and vendoring a `package-lock.json`
- Move the Released Notes sections from 24.05 to 24.11

**Review by SuperSandro2000 — APPROVED** [review]
LGTM

**Comment by poperigby** [comment]
Maybe it would be a good idea to figure out the problem with parsing the template from Nix before merging this?

**Comment by poperigby** [comment]
It might also be a good idea to set an list option in the `secrets` attribute set, `jwksFile`, so Nix can handle the templating for you?

**Comment by nicomem** [comment]
Changes:
- Updated to 4.38.9
- Changed the pnpm version used from 8 to 9 (same as remote)

For the jwks, I currently do not have much time to experiment and try to find a solution.
And since this is not a blocking issue but more of a QoL improvement, this could be done in the future.
But if someone has a good and secure solution to set the jwks from the NixOS module config, I could add it to this PR.

**Comment by poperigby** [comment]
@nicomem Fair enough. I opened an issue for being able to convert the Go templates to Nix: https://github.com/NixOS/nixpkgs/issues/319716

File: nixos/modules/services/security/authelia.nix
Comment (bendlas) [inline nixos/modules/services/security/authelia.nix:378]: Right now, this is throwing a way more confusing error message:

`error: The option `services.authelia.instances.<name>.secrets.oidcIssuerPrivateKeyFile' does not exist. Definition values:`

If we cannot use the deprecation facilities, we'll still have to leave the option in, in order for the user to ever see this assertion message.

File: nixos/modules/services/security/authelia.nix
Comment (bendlas) [inline nixos/modules/services/security/authelia.nix:80]: Probably better to leave in the defunct option, than to remove the dead assertion.

I think we can mark it as `internal`, so that it doesn't show up in the option reference any more.

**Review by bendlas — CHANGES_REQUESTED** [review]

**Comment by bendlas** [comment]
Also, would it be much trouble, to keep accepting the deprecated `oidcIssuerPrivateKeyFile`, while authelia keeps accepting it (so for the 4.38 release cycle)? That would give nixos authelia users the same upgrade window, that authelia has.

**Comment by bendlas** [comment]
Also also: Couldn't we keep the `secrets.oidcIssuerPrivateKeyFile` option and do whatever templating shenanigans authelia needs to support this, inside the module?

i.e. use the recommended template yml as a `writeText`, until https://github.com/NixOS/nixpkgs/issues/319716 is resolved)?

**Comment by hatch01** [comment]
Hi, I don't know if it could help but I use this pr to overlay the authelia package and since the last push, I get a hash mismatch:
```console
error: hash mismatch in fixed-output derivation '/nix/store/aq4s7c04kkjl5zbsi6q8iv3dv5i7zprn-authelia-web-pnpm-deps.drv':
         specified: sha256-xhn8+eBf0pF1ujglIoBmj0E1+B3YngEESb/SZlR4Ugk=
            got:    sha256-2kABLWeelTml1tcxcgWLHWv/SNqwGxJB+VExusnd+KI=
error: 1 dependencies of derivation '/nix/store/7zhha0394q4jkf77m5wv11rvf63a04wc-authelia-web-4.38.9.drv' failed to build
error: 1 dependencies of derivation '/nix/store/7y62sqb1ab00pmar1n47k5srp9vvf1rr-authelia-4.38.9.drv' failed to build
error: 1 dependencies of derivation '/nix/store/2j98ynd7vwanaal5772yyfk4b9y899w6-unit-authelia-main.service.drv' failed to build
```

And just a question but is there an easy method to use the new service from this PR like overlays ?

**Comment by alapshin** [comment]
> And just a question but is there an easy method to use the new service from this PR like overlays ?

I do it this way.

```
{ lib
, pkgs
, config
, inputs
, ...
}: {
  disabledModules = [
    "services/security/authelia.nix"
  ];
  imports = [
    "${inputs.authelia}/nixos/modules/services/security/authelia.nix"
  ];
   authelia.instances."main" = {
      enable = true;
      package = pkgs.prs.authelia; 
  };

}
```

Where `inputs.authelia` is a flake input definition
```    
inputs.authelia.url = "github:nicomem/nixpkgs/authelia-4.38";
```

and `pkgs.prs.authelia` is a new package from overlay
```
  pr-packages = final: _prev: {
    prs = import inputs.authelia {
      system = final.system;
    };
  };
```

**Comment by poperigby** [comment]
You can also just do `inputs.authelia.legacyPackages.${system}.authelia` for the `package` field and skip using an overlay.

**Comment by bendlas** [comment]
@hatch01 I've got this in my `flake.nix` (switched back to my old PR for now):

```nix
  outputs = { self, ... }@inputs
  : let
    nixpkgs = import ./patched-nixpkgs.nix self inputs.nixpkgs "x86_64-linux" [
      ## authelia 4.38
      # ## TODO https://github.com/NixOS/nixpkgs/pull/299309
      { url = "https://github.com/NixOS/nixpkgs/pull/317164.patch";
        hash = "sha256-x7ba8hYxvAfmUln5KSZDcp2AQ4BzL77yccIAR8Pu0sk="; }
      # ## authelia openapi
      # { url = "https://github.com/NixOS/nixpkgs/pull/320645.patch";
      #   hash = "sha256-dmQIdd7EdyRbiZ41DJZrZ/d/2A5wtAkaocHQu1zIuWM="; }
    ];
   in ....
```

where `patched-nixpkgs.nix` is

```nix
## https://discourse.nixos.org/t/proper-way-of-applying-patch-to-system-managed-via-flake/21073/26
self: nixpkgs: system: fetchpatchDescriptors: let

  pkgs-boot = import nixpkgs { inherit system; };

  patches = map
    (dsc: pkgs-boot.fetchpatch dsc)
    fetchpatchDescriptors;

  nixpkgs-patched =
    pkgs-boot.applyPatches {
      name = "nixpkgs-patched";
      src = nixpkgs;
      inherit patches;
    };

in ((import "${nixpkgs-patched}/flake.nix").outputs { inherit self; })
   // {
     sourceInfo.outPath = nixpkgs-patched;
   }
```

**Comment by nicomem** [comment]
@bendlas
- On the error message:
  - Good catch! I readded the option and put an `internal = true;`
- On accepting `oidcIssuerPrivateKeyFile`:
  - I am quite against writing the file content directly into the YAML file: secrets should not be put into the nix store as it is readable by everyone
  - #319716 is the work to do to be able to set this secret securely from the module. From my understanding, the only secure workaround is to do something like what was said previously: [comment 1](https://github.com/NixOS/nixpkgs/pull/299309#issuecomment-2045898077), [comment 2](https://github.com/NixOS/nixpkgs/pull/299309#issuecomment-2108501250)

@hatch01
- Good catch, my bad. Fixed it, `nixos-rebuild` now works on my machine

**Comment by bendlas** [comment]
>   * I am quite against writing the file content directly into the YAML file: secrets should not be put into the nix store as it is readable by everyone

@nicomem are you sure, that this is what would be happening? I'm under the impression, that the yaml file could be something like

```nix
writeText "oidc-jwks.yaml" ''
  identity_providers:
    oidc:
      jwks:
        - key: {{ secret "${oidcIssuerPrivateKeyFile}" | mindent 10 "|" | msquote }}
```

And that would be resolved by authelia at load time, as per https://www.authelia.com/reference/guides/templating/
No secret going to the store. Am I mistaken here?

What would https://github.com/NixOS/nixpkgs/issues/319716 (effectively a DSL in front of the `writeText`) change about what goes to the store?

**Comment by nicomem** [comment]
@bendlas Sorry I misunderstood what you were trying to say.
That is a good idea, which should probably work! I'll try doing that

**Comment by nicomem** [comment]
The solution seems to work well! I added this modification to the PR

File: nixos/doc/manual/release-notes/rl-2411.section.md
Comment (AndersonTorres) [inline nixos/doc/manual/release-notes/rl-2411.section.md:13]:
```suggestion
  This release also deprecates some configuration keys, which are likely to be removed in future version 5.0, but they are still supported and expected to be working in the current version.
```

**Review by AndersonTorres — APPROVED** [review]

File: nixos/modules/services/security/authelia.nix
Comment (bendlas) [inline nixos/modules/services/security/authelia.nix:386]: Haven't tested this, but wouldn't the `host` and `port` options have the same issue of never getting to this assertion ....

File: nixos/modules/services/security/authelia.nix
Comment (bendlas) [inline nixos/modules/services/security/authelia.nix:159]: ... when the `host` and `port` options are removed?

File: nixos/doc/manual/release-notes/rl-2411.section.md
Comment (bendlas) [reply nixos/doc/manual/release-notes/rl-2411.section.md:13]: > This release also deprecates some configuration keys ... but they are still supported and expected to be working in the current version

which options would this apply to specifically?

Comment (nicomem) [reply nixos/doc/manual/release-notes/rl-2411.section.md:13]: As per [the release blog post](https://www.authelia.com/blog/4.38-release-notes/):

> There are some changes in this release which deprecate older configurations, you will get a warning about these deprecations as it’s likely in version v5.0.0 we’ll remove support for them, however if a log message for a configuration is a warning then it’s just a warning, and can fairly safely be ignored for now.

The blog post does not list the configuration keys that are deprecated.

File: nixos/modules/services/security/authelia.nix
Comment (nicomem) [reply nixos/modules/services/security/authelia.nix:386]: `host` and `port` do not have the same problem: if I try to set `settings.server.host = "foo"`, I get the assertion error.

I think this is because `host` and `port` are part of the `settings` option which allows setting even values that are not defined in the module (as the entire `settings` value is then formatted as the main YAML config file).

**Comment by nicomem** [comment]
Changes: `depreciates` -> `deprecates`

File: nixos/modules/services/security/authelia.nix
Comment (bendlas) [reply nixos/modules/services/security/authelia.nix:386]: Ah, ok, got it.

Comment (bendlas) [inline nixos/modules/services/security/authelia.nix:386]: So I'll again ask you to please downgrade the assertion to a warning.

**Review by bendlas — CHANGES_REQUESTED** [review]
Found more problems during testing: `error occurred performing deprecation mapping for keys 'server.host', 'server.port', and 'server.path' to new key server.address: the new key already exists with value 'unix:///run/authelia/http' but the deprecated keys and the new key can't both be configured`

- According to the error message, the `address` option not just supersedes `host` and `port`, but also the `path` option.

- ~I am using the the `path` configuration option. Right now, I'm not sure how to combine that into the `address` option, together with a unix socket path. This is not a problem to necessarily be solved as part of this PR - I will ask about it upstream -~ but it is an example for why it's preferrable to not put additional restrictions on what's accepted upstream. **EDIT**: found out, you can pass a `?path=...` argument to the unix socket address.

~Second Error: `identity_providers: oidc: option `jwks` must not be configured at the same time as 'issuer_private_key' or 'issuer_certificate_chain'`~ **EDIT**: solved it. I could update my cert chain config. Will contribute this.

- This one looks, like it should be fixed in my config. I'll check on how to do that.

File: nixos/modules/services/security/authelia.nix
Comment (bendlas) [reply nixos/modules/services/security/authelia.nix:386]: One more reason: With the `path` configured in a unix socket `address` , it seems that it's not picked up for consent screens, so I'll have to stay on the old config, until this can be resolved upstream ...

**Comment by bendlas** [comment]
So, I managed to get everything running with this PR now, but I feel like the assertion doesn't cut it. To summarize:

- Using `address` also precludes using `path`: We could add this to the assertion, but I'd recommend removing it outright
- Doing an assertion precludes using the still valid (if deprecated) options: I recommend downgrading to a warning, or removing the check outright (since we won't be validating all the authelia options anyway)

File: nixos/modules/services/security/authelia.nix
Comment (bendlas) [reply nixos/modules/services/security/authelia.nix:386]: summarizing in comment below

**Comment by nicomem** [comment]
@bendlas I removed the assertion because as you said, we won't validate all the options.

Another thing that I thought is about the default value of the `server.address` setting:
- Defining a default value in the module will break the config of those that are defining `host`, `port` or `path`
- But most NixOS modules define a default value for the host+port/address ; and when the deprecated settings will be outright removed, the config will stop to work in any case

**Review by bendlas — APPROVED** [review]
Excellent!

**Comment by hatch01** [comment]
Hi, are you using the log.file_path parameter.
Because in my installation, setting it to "/var/log/authelia/authelia.log" just make authelia crashing:
```console
juil. 07 16:17:57 nixos authelia[18380]: time="2024-07-07T16:17:57+02:00" level=fatal msg="Cannot configure logger: open /var/log/authelia/authelia.log: read-only file system" stack="github.com/authelia/authelia/v4/internal/commands/root.go:69 (*CmdCtx).RootRunE\ngithub.com/spf13/cobra@v1.8.1/command.go:985                 (*Command).execute\ngithub.com/spf13/cobra@v1.8.1/command.go:1117                (*Command).ExecuteC\ngithub.com/spf13/cobra@v1.8.1/command.go:1041                (*Command).Execute\ngithub.com/authelia/authelia/v4/cmd/authelia/main.go:10      main\nruntime/internal/atomic/types.go:194                         (*Uint32).Load\nruntime/asm_amd64.s:1695                                     goexit"
```
I don't know if it is caused by the update or if it was broken before.

Ps:
I can obviously write in the file using `sudo -u authelia echo test > /var/log/authelia/authelia.log` and read it in the same way.

File: nixos/modules/services/security/authelia.nix
Comment (hatch01) [reply nixos/modules/services/security/authelia.nix:161]: I tried and did not manage to make it working.
Do you have a complete authelia config example available on github ?

**Comment by nicomem** [comment]
@hatch01 
> Hi, are you using the log.file_path parameter.
> Because in my installation, setting it to "/var/log/authelia/authelia.log" just make authelia crashing:

The authelia instance `services.authelia.<name>` declares in its systemd settings the state directory `/var/lib/authelia-<name>` (and not `/var/lib/authelia`). Changing the  `log.file_path` to point to the correct directory should probably solve your problem.

File: nixos/modules/services/security/authelia.nix
Comment (nicomem) [reply nixos/modules/services/security/authelia.nix:160]: I could not find a way to make this work:

The first problem is that there can be multiple authelia instances, some using the old settings, and others using the new one -> The default value being applied to all, there will be problems.

If we suppose that there is only one instance, after fixing the condition (the settings are accessed with `cfg.instances.<name>.settings.server.{host,port,path}`), this still generates the `server.address` setting in the YAML file, with the value `null`.

I tried to instead conditionnaly set the `default` value by doing this:
```nix
address = mkOption ({
    type = types.str;

    example = "unix:///var/run/authelia.sock";
    description = "The address to listen on.";
  } // lib.optionalAttrs
    (
      let
        instance0 = builtins.elemAt (builtins.attrValues cfg.instances) 0;
      in
      if (instance0.settings.server.host != null) ||
        (instance0.settings.server.port != null) ||
        (instance0.settings.server.path != null) then
        warn "Please replace services.authelia.setings.{host,port,path} with services.authelia.settings.address, before release 4.39" false
      else true
    )
    {
      default = "tcp://:9091/";
    });
};
```
, but nix encounters an infinite recursion...

File: nixos/modules/services/security/authelia.nix
Comment (nicomem) [reply nixos/modules/services/security/authelia.nix]: Modification applied

**Comment by hatch01** [comment]
> @hatch01
> 
> > Hi, are you using the log.file_path parameter.
> > Because in my installation, setting it to "/var/log/authelia/authelia.log" just make authelia crashing:
> 
> The authelia instance `services.authelia.<name>` declares in its systemd settings the state directory `/var/lib/authelia-<name>` (and not `/var/lib/authelia`). Changing the `log.file_path` to point to the correct directory should probably solve your problem.

Sorry I forgot to precise my config : https://github.com/hatch01/server-flake/blob/354a795dff56383e466ee414652a0e2441ccf2f3/apps/authelia.nix
I already set file_path which is the settings which causes the problem.

**Comment by nicomem** [comment]
Sorry if my explanation was not clear, but in your config, you are setting the `log.file_path` to a file in the `/var/lib/authelia` directory whereas the systemd service uses `/var/lib/authelia-<name>` (which in your case would be `authelia-main`).

Try setting `log.file_path` to `/var/lib/authelia-${autheliaInstance}/authelia.log`.

If for some reason you still want/need to use the `/var/lib/authelia` directory, you will probably need to change the service's systemd settings to allow read-write access to it.

File: nixos/modules/services/security/authelia.nix
Comment (bendlas) [reply nixos/modules/services/security/authelia.nix]: > I tried and did not manage to make it working. Do you have a complete authelia config example available on github ?

What are you having your issue with @hatch01?

Have you had a working configuration with authelia on a subpath before?

I do have a working configuration. It's part of a larger approach, that I'm looking at open-sourcing real soon (tm). If you'd be interested in having a look at that, possibly helping out, I'm happy to share privately. Feel free to DM.

File: nixos/modules/services/security/authelia.nix
Comment (bendlas) [reply nixos/modules/services/security/authelia.nix:160]: Hm, yeah, I think the infinite recursion comes from `builtins.elemAt` and/or `builtins.attrValues` being eager.

Also it seems like that snippet would set the default for each authelia instance, based on the setting of the (arbitrary) first one. I don't think, that's what we want to be doing.

Try taking a `config` argument in the function for `autheliaOpts`, see comment below.

File: nixos/modules/services/security/authelia.nix
Comment (bendlas) [inline nixos/modules/services/security/authelia.nix:12]:
```suggestion
  autheliaOpts = with lib; { name, config, ... }: {
```
This `config` will be scoped to per-instance, because `autheliaOpts` is passed to `types.submodule`.

That way, it should be possible to do warnings per affected instance.

Comment (nicomem) [reply nixos/modules/services/security/authelia.nix:12]: Using `config` here allows for accessing the instance configuration (thank you for the tip), however I still get the infinite recursion :/

I think that it is because I try to lazily define `settings` (i.e. the `address` option of the submodule `settings`) while trying to access the "final state" of `settings` (i.e. the `host`/`port`/`path`).

When I set the condition to something outside `settings` (e.g. `config.enable`), it works.
It is when I try to use something in `settings` that it breaks.

Also, when using eager functions (e.g. `lib.optionalString`), it compiles, but the key is still defined in the YAML configuration file (`address: ''`), and Authelia yells about defining both old and new keys.

Comment (bendlas) [reply nixos/modules/services/security/authelia.nix:12]: Hm :/ I'd have to take a look to say more, but I guess last resort would be a top-level `mapAttrsToList`, to generate `config.warnings` manually ....

**Comment by dimakow** [comment]
Hi folks,

may I ask if you have any ETA in mind for this PR ?

Regards

**Comment by nicomem** [comment]
Finally found a way to make the default value of `address` work with `host`/`port`/`path`!!!

Instead of trying to modify whether we set the default value of `address` depending on whether the others have been set, we can instead take the `instance.settings` attribute set and remove the `address` value if the others have been set, before passing it to the YAML formatter.

This seems to work well:
- Setting any of the old settings displays a warning, and `address` does not appear in the config file
  - The warning is per-instance (i.e. only the instances that use the old settings will trigger a warning, one for each)
  - The warning indicates which instance defines the old settings (e.g. `Please replace services.authelia.instances.main.settings.{host,port,path} with services.authelia.instances.main.settings.address, before release 5.0.0`)
- Not specifying the old settings, and either setting or leaving as default `address` does not display any warning, and the value is set in the config file

This now means that this PR now does not bring any breaking change!

Thanks again @bendlas for all the comments and ideas to make this work

**Comment by bendlas** [comment]
Tested and works. Thanks @nicomem, for all your work and being receptive to feedback!

I would merge this right now, except it feels like this should be something done by package maintainers. @06kellyjac and @dit7ya don't seem to have participated so far. Let's wait for the weekend, if they get around, otherwise I'll merge.

> may I ask if you have any ETA in mind for this PR ?

@dimakow Generally, there are no ETAs, as NixOS is volunteer work. A more productive way to ask something like this, is to pick the PR onto your branch, try it out, and provide feedback based on your experience, or simply ask if there is any way to help move this along.

**Review by 06kellyjac — APPROVED** [review]
image


image

I've not seen it necessary to repeat valid suggestions already posted on quite an active PR.

I've not requested the commit-bit for nixpkgs so I can't merge.

LGTM I don't mind merging now or waiting for @dit7ya

**Comment by bendlas** [comment]
@06kellyjac my bad, sorry. Thanks for approving. I'll merge.

**Pull request merged**
